CPT & TUNL Protocol Updates

FilipKosel opened the pull request:
Underlying CPT protocol updated to align with mouse version (e.g., 0.33/0.67 target probability; minimum 80% hit rate and maximum 40% false alarm rate for staircasing; remove Noise and Blur difficulty scaling and probes) and reflect changes from recent development meetings (e.g., remove limited hold difficulty scaling; increase duration of Flanker probe; set Target Probability probes to start with low target probability).

Expanded list of CPT changes:

- Changed stage names to better reflect what happens
        - Added 'Staircase' to staircased stages and 'Fixed' to stages with fixed parameters
        - Used 'Difficulty' in place of 'Scaling' for stages that adjust difficulty; used 'Probe' for probe stages

- Removed limited hold difficulty stage and blur/noise stages
        - Will be re-added in future update to align with new staircasing methods

- Trial lists and associated variables separated into two categories: "base" (i.e., standard 0.33/0.67 as used in mice) and "hilo" (i.e., high/low target probabilities for TarProb and SART stages)
        - Configuration includes target probabilities, number of trials in the list, and max run for constrained shuffle for both “base” and “hilo” categories

- Staircasing updated
        - Staircasing based on minimum 80% hit rate and maximum 40% false alarm rate over 15 trials (can be set in configuration; same criteria as in mice)
        - Once staircasing is triggered, it continues under the following conditions until an end condition is met
                - Increases only occur during correct rejections (i.e., can differentiate target/non-target)
                - Decreases only occur during false alarms (i.e., cannot differentiate target/non-target)
        - Similarity staircasing end conditions:
                - Staircasing changes direction (i.e., criteria met then failed; criteria failed then met)
                - Staircasing increases and a correct rejection is made at the maximum similarity value (i.e., no further increase possible)
                - Staircasing decreases and reaches the minimum (i.e., no further decrease possible)
        - Stimulus duration staircasing uses binary search to determine minimum stimulus duration

- Other changes
        - Separate hit rate/false alarm rate tracking lists created (also used for accuracy for end of stage results)
        - Creation of trial lists simplified
                - Trial lists now based on the minimum number of target and non-target trials (determined by trial list length and target probability provided in configuration file)
                - Minimum target/non-target trials also used for staircasing criteria (i.e., staircasing only occurs after both minimums met)
        - Separate trial lists used for each Flanker substage to ensure equal number of target and non-target presentations for each Flanker type
        - Flanker variables now set separately from other trials in trial_contingency to ensure appropriate trial lists are used
        - Separate function created to present tutorial instruction text instead of video to keep same initial screens
        - In block contingency, parameters for upcoming stage initially set to default values; only necessary values changed
        - For stimulus duration staircasing probe, initial stimulus duration set to mean response time for hits (i.e., latency between stimulus onset and finger lift from hold button); if no response times in list, set initial stimulus duration to 1/2 of current limited hold value
                - Response times tracked for hits; check response_latency is not np.nan before appending to list.
        - Instruction file updated to include tutorial text for no video condition
        - Configuration files updated to reflect changes
        - start_stage_screen updated to unbind iti_start and premature_response from hold_button to prevent stimulus/feedback from appearing over results screen
        - stage_screen_end updated to bind iti_start to hold_button


Underlying TUNL protocol updated to improve staircasing of combo probe; staircasing now based on 7/8 correct responses; delay staircased separately for each separation using binary search method; separations and minimum and maximum delays set in configuration file; previous space and delay probes removed for now.

Expanded list of TUNL changes:

- Staircasing for combined (separation and delay) probe revised
        - Staircasing criteria changed to require 7/8 hits to increase staircase, else decreased
        - To reduce time, staircase decreases as soon as two incorrect responses made
        - Response tracking reset after every increase/decrease (maximum 8 trials)
        - Staircasing uses binary search
                - Minimum and maximum delay values set in configuration file
                - Initial delay set to mean of minimum and maximum
                - If staircasing up, minimum delay set to current delay (at which criteria were met)
                - If staircasing down, maximum delay set to current delay (at which criteria were failed)
                - Mean of minimum and maximum delay taken for next current delay
                - This continues until minimum and maximum values converge within delay resolution value (set in configuration file)
                - Final value for each separation is current minimum delay (i.e., highest value at which staircasing criteria were met)
                - Minimum delay at one separation used as maximum delay for next harder separation
                        - i.e., if minimum (final) delay at which criteria met for 2 units of separation is 12s, the starting maximum delay for 1 unit of separation is 12s regardless of what delay limits given in configuration file

- Previous space and delay probes removed to clean up code; can be readded in the future

- Variables for tracking combo probe values revised and cleaned up
        - Now uses dictionaries with separations as keys for tracking multiple values

- Results screen now displays maximum delay at which criteria met for each separation individually.

- "present_tutorial_text" function added as alternative to tutorial video; just presents text instructions

- Start protocol function now triggers block contingency at its end to properly set values and initiate trial contingency

- size_hint and pos_hint added whenever instruction_label and instruction_button displayed due to incorrect size/location behaviour on Surface device (unknown origin of issue)

- Block start and trial end times now set at end of block_contingency function (block_end function not called)

- Block duration significantly increased to allow for all staircasing trials to be completed

- Configuration and configuration-backup files adjusted to match

KNOWN ISSUES:

- Delay and tutorial videos do not fill screen on Surface device

- Delay and tutorial videos overlap buttons (hold button for delay video; restart tutorial and start task buttons for tutorial video)

- Stimulus sizing wildly different across devices

* These issues are likely caused by screen resolution settings

- protocol_floatlayout.add_button_event and add_text_event cause errors indicating that too many arguments passed to functions; issues not present for other add_*_event calls.

dpalmer9 commented:
Error in Flanker probe:

On a test run with 3 FA, I see an accuracy metric of 100%.

To confirm, are you making accuracy the HR? How are you calcualting it?

FilipKosel commented:
Summary of changes for most recent commit (CPT/TUNL Output/Result Screen Edits)


CPT:

- Similarity SSIM value now displays for all stages if "Similarity" is in the stage list
- Results screen (start_stage_screen) wording adjusted for clarity and simplicity
        - Removed reference to frames from Stimulus Duration stage results for simplicity
        - Altered generic (accuracy) results to specify percentage of responses to target images and number of responses to non-target images
- Added/updated definition for trial_outcome 7 (trial abort; lift and press hold button during stimulus display)
- For all outcome functions, adjusted trial_outcome comments to reflect only possible outcomes for that function
- Changed "CurrentBlockTrial" column to "BlockTrial" column in data output for consistency with other columns
- Removed block_hits variable (unused)
- Removed 'stimulus_on_screen = False" from stimulus_end function
        - During stimulus duration probe, this would allow premature_response function to be triggered during limited hold period
- Unscheduled stimulus_end and center_notpressed event triggers in response_cancelled function
        - If response_cancelled function triggered during stimulus presentation period of stimulus duration probe (before limited hold mask), this could cause the limited hold mask to appear even after the stimulus was removed from the screen and the trial was cancelled
        - center_notpressed triggered at end of response_cancelled function anyway
- Changed trial contingency if/else statement checking if block trials exceeded to check if block trials greater than limit
        - Was previously greater than or equal to; would cause block end to be triggered one trial early

- NOTE: Stimulus can still appear after stage results screen appears if screen held at hold button location even after hold button removed
        - ITI and stimuli will continue to cycle over stage results as long as screen held

TUNL:
- Adjusted columns for data output to align with CPT
- Created contingency and trial_outcome variables and set defaults to np.nan
        - Trial contingencies and outcomes align with CPT
        - All functions that impact trial outcome now have comments indicating possible and specifically assigned outcome values
        - Trial outcomes are as follows (based on CPT):
                - 0: Premature response (lift during ITI)
                - 1: Hit (target response)
                - 2: Miss (no response)
                - 3: False alarm (non-target response)
                - 7: Cue abort (lift finger during cue presentation)
                - 8: Delay abort (lift finger during delay presentation)
- Changed local trial_coord variable name in generate_trial_pos_sep to combined_coord to distinguish from global self.trial_coord

dpalmer9 commented:
> * NOTE: Stimulus can still appear after stage results screen appears if screen held at hold button location even after hold button removed
>   - ITI and stimuli will continue to cycle over stage results as long as screen he

Issue resolved by forcing a state change of self.hold_button.

self.hold_button.disabled = True <-- Disables the buttons behaviour and stops any activity
self.hold_button.state = 'normal' <-- Resets the current stage of the widget to not pressed

....
self.hold_button.disabled = False <-- Safely re-enable the widget
